### PR TITLE
ForwardMode phi

### PR DESCRIPTION
**Pull request by tgymnich**
Added derivative of phi for forward mode

### CODE REVIEW

File: enzyme/Enzyme/AdjointGenerator.h [hunk 1]
```diff
@@ -395,6 +395,8 @@ class AdjointGenerator
           auto diff = Builder2.CreateLoad(
               gutils->invertPointerM(I.getOperand(0), Builder2));
           setDiffe(&I, diff, Builder2);
+        } else {
+          setDiffe(&I, Constant::getNullValue(I.getType()), Builder2);
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:399]: Why is this necessary, there shouldn't be a diffe set for inactive values?

File: enzyme/Enzyme/AdjointGenerator.h [hunk 2]
```diff
@@ -1466,6 +1511,8 @@ class AdjointGenerator
 
     Type *addingType = BO.getType();
 
+    setDiffe(&BO, Constant::getNullValue(BO.getType()), Builder2);
```
Comment (wsmoses) [inline enzyme/Enzyme/AdjointGenerator.h:1514]: Why do we need this? If this op is constant we should never set it, and if it is active we should always set it to the computed derivative?

File: enzyme/Enzyme/AdjointGenerator.h (hunk 1 above)
Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h:399]: This was not supposed to be in this PR.

File: enzyme/Enzyme/AdjointGenerator.h (hunk 2 above)
Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h:1514]: Because I am using `addToDiffe` I need to "reset" variables that are used in loops. If I stop using `addToDiffe` I can get rid of this.

Comment (wsmoses) [reply enzyme/Enzyme/AdjointGenerator.h:1512]: We should just use setDiffe for forward mode (and arguably can assert in addToDiffe if in forwardMode).

Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h:1512]: Already working on this 😄

Comment (wsmoses) [reply enzyme/Enzyme/AdjointGenerator.h:1557]: Bump at potentially removing this?

**Comment by wsmoses** [comment]
Besides the unnecessary setDiffe, LGTM.

**Review by wsmoses — APPROVED** [review]

File: enzyme/Enzyme/AdjointGenerator.h (hunk 2 above)
Comment (tgymnich) [reply enzyme/Enzyme/AdjointGenerator.h:1557]: completely forgot to remove this 🙈
